src/main/java/com/pinterest/secor/uploader/HadoopS3UploadManager.java

```diff
@@ -60,12 +61,33 @@ public HadoopS3UploadManager(SecorConfig config) {
         LOG.info("uploading file {} to {}", localLogFilename, logFileName);
 
         final Future<?> f = executor.submit(new Runnable() {
+            final int totalRetries = Integer.parseInt(mConfig.getUploaderRetries());
+            final long backoffMillis = Long.parseLong(mConfig.getUploaderRetryBackoffMillis());
             @Override
             public void run() {
-                try {
-                    FileUtil.moveToCloud(localLogFilename, logFileName);
-                } catch (IOException e) {
-                    throw new RuntimeException(e);
+                int retriesLeft = totalRetries;
+                boolean success = false;
+                while (retriesLeft > 0 && !success) {
+                    try {
+                        FileUtil.moveToCloud(localLogFilename, logFileName);
+                        success = true;
+                    } catch (IOException e) {
+                        success = false;
+                        retriesLeft--;
+                        if (retriesLeft < 0)
```

Comment:
Looks to me that counting might be off by one.  When retriesLeft gets to 0, it will not trigger this if statement, instead it will fall off the while loop, thus the last exception is not thrown.

Usually people record the last exception in a variable during the loop, have the exception thrown after the while loop if success = false.  This way you don't have to match the counter between the while loop and if statement.